### DISCUSSION TITLE
Need general help installing and running with websockets

### DISCUSSION THREAD

**Discussion by ttonyh** [cagnulein/qdomyos-zwift#1461] [General]
@cagnulein I have a few questions regarding running QZ with websockets and compiling on RPI. Thanks in advance for your help:

1. Do I need to compile/install Qt Httpserver (outlined in 40_web_socket_api.md) to get websockets working? I can get QZ running on my Raspberry Pi Zero W and it can connect to my Echelon EX-5 Connect bike but I cannot connect to QZ through any port. I don't need a GUI interface, just the API.

2. I compiled with RPI Zero W but took about 6 hours. Can you recommend a faster way other than using another Raspberry Pi (which are hard to get these days). Maybe using qemu; if you've tried that - a setup that allows for compatibility and being able to maximize the host resources would be ideal.

Thank you!

**Comment by cagnulein** [comment]
@ttonyh 
1. yes :)
2. cross compiling from your pc is the only way to achieve this. anyway 6 hours is only for the first time when you need to build everything :)

**Reply by ttonyh** [reply]
@cagnulein Thank you for your fast response; although QZ was very easy to compile, the websocket part (QDomyos-Zwift etc.) did not work for me. I wonder if the instructions in "40_web_socket_api.md" are out of date? I tried it a couple of weeks ago and didn't save the output, so no logs to show.

**Reply by cagnulein** [reply]
@ttonyh they seem ok to me. of course, in order to help you, i need the exact error. let me know!

**Comment by ttonyh** [comment]
@cagnulein I have it working somewhat in a VM via Virtualbox on Debian 11 Bullseye. I'm able to connect for a few minutes but then get a segmentation fault. I tried installing the Qt Httpserver, which seemed to install fine, but I can never connect. Is there a place where logs are written to disk? I'm unable to direct the output logs to a file.

**Comment by cagnulein** [comment]
@ttonyh the debug log from qz? just enable the debug log in the experimental settings and you will find the file in the same binary directory

**Comment by ttonyh** [comment]
Is 15.12.12 still the recommended version of QT to use or should I compile a more recent of v15?

**Reply by cagnulein** [reply]
@ttonyh what is the 15.12.12? qz uses qt v5.15

**Reply by ttonyh** [reply]
Sorry, typo. I meant 5.12.12, referenced here in "40_web_socket_api.md".

**Reply by cagnulein** [reply]
use the 5.15 :)

**Reply by ttonyh** [reply]
What flags do you recommend I use for the QT configuration?

**Reply by cagnulein** [reply]
it's been a while since i recompiled the qz last time, you can start from the flags that you can find in thw 40_web_socket_api.cmd file. Anyway building the QT from scratch is not required to have the web socket. Why are you doing this?

**Reply by ttonyh** [reply]
Yes, my only goal is to get websockets working. I would love it if I didn't have to compile QT. What do you suggest?

**Reply by cagnulein** [reply]
as far as I know now raspberry has the qt 5.15 so you should just use the builin qt library and compile only the httpserver

**Reply by ttonyh** [reply]
Compiling the QtHttpServer module I get the following error; I tried searching but there's not a lot of info out there on it for QT5... is it a QT6 specific feature: "websockets-private"

```
Project ERROR: Unknown module(s) in QT: websockets-private
```

**Reply by cagnulein** [reply]
i'm building the httpserver on the CI here https://github.com/cagnulein/qdomyos-zwift/blob/7ba1986929f488acc587fa5ea944dcf6e444a1d0/.github/workflows/main.yml#L90

are you following the same workflow?

**Reply by cagnulein** [reply]
also check this https://forums.raspberrypi.com/viewtopic.php?t=206082

**Reply by ttonyh** [reply]
Thank you; I will try to your suggestions.

**Reply by gungun974** [reply]
Hello then it was me indeed who wrote this tutorial or so on how to operate the Websocket for my Raspberry Pi Zero.

***And indeed I am sorry to say it but my tutorial is well outdated.***

And the worst in this story is that I tried before I saw this conversation to recompile the project without any success for the moment. 👀 
*(Yay I’m back me too at square one !)*

**Reply by gungun974** [reply]
But anyway, I'm going to try now to give you what I've already tried and succeeded in doing xD

As I was saying in my tutorial, you need to compile a version of QT to have a module that I suppose is not compiled in the official Bullseye repositories. (Actually, I'm not sure).

Thanks to my written sources, I quickly found the https://www.tal.org/ website and a step-by-step tutorial to compile a more recent version of QT (5.15.7).

However, to compile QT, it took me about 5 hours on a Raspberry Pi 4 and A LOT of package installations from development which I can't really say which one was necessary.

But once I managed to install this version of QT, I was able to compile the qthttpserver module without any error messages with my instructions and the same for Qdoysmos.

After that, I noticed that my Raspberry Pi, when I just launched Qdoysmos, didn't open any port despite the working Bluetooth and then by inspecting the code step by step that leads to the C++ class that I want QT to load, I realized that I had to write "-qml" in the CLI to load the Home page that will load this file afterwards.

Now I have a problem, though. My C++ program segment faults for no reason when the program loads the Home page.

**Reply by ttonyh** [reply]
I have a new discussion linked below, working with @cagnulein on getting this to work. My experience currently is, that QZ is easy to compile with no problem, the only thing I can't figure out is how to get the websockets to work. Robert has mentioned that compiling QT5 is unnecessary; which is fine but the main problem is QtHttpServer. It always gives me the error I mentioned at top of this thread. I don't think the solution is to remove "websocket-private" from the code. Is this module only for when running on Windows? The CI actions don't have any reference to it except for the Windows build. (https://github.blog/changelog/2022-01-11-github-actions-jobs-running-on-windows-latest-are-now-running-on-windows-server-2022/)

https://github.com/cagnulein/qdomyos-zwift/discussions/1461

**Reply by gungun974** [reply]
Anyways, now I suppose I probably have a problem with the way I compiled the qhttpserver module and with the CI of cagnulien, maybe I'll be able to better build and eliminate this segment faults.

If anyone here, however, would be very kind and patient to explain to me how with GDB I can find out where the program is segfaulting, I would be grateful for this help.

**Reply by gungun974** [reply]
> I have a new discussion linked below, working with @cagnulein on getting this to work. My experience currently is, that QZ is easy to compile with no problem, the only thing I can't figure out is how to get the websockets to work. Robert has mentioned that compiling QT5 is unnecessary; which is fine but the main problem is QtHttpServer. It always gives me the error I mentioned at top of this thread. I don't think the solution is to remove "websocket-private" from the code. Is this module only for when running on Windows? The CI actions don't have any reference to it except for the Windows build. (https://github.blog/changelog/2022-01-11-github-actions-jobs-running-on-windows-latest-are-now-running-on-windows-server-2022/)
> 
> #1461

Personally, I have no idea why, but if your error indicates that files do not exist in the QtHttpServer project that you clone and compile with the default version of QT for Raspberry Pi, I can assure you that I had the same problem. Once I broke my head compiling a version of QT from A to Z, I no longer have this error at all.

**Reply by ttonyh** [reply]
I got QZ by itself to compile, run and connect perfectly on the PI Zero W. Took about 6 hours but no issues. However I can't do anything with it because no websockets. Before I go back to compiling on PI, I an trying everything out in Docker first. But the problems are as I described above and in the other discussion. What makes working with qthttpserver even more difficult is that the repo has zero documentation.

**Reply by gungun974** [reply]
> I got QZ by itself to compile, run and connect perfectly on the PI Zero W. Took about 6 hours but no issues. However I can't do anything with it because no websockets. Before I go back to compiling on PI, I an trying everything out in Docker first. But the problems are as I described above and in the other discussion. What makes working with qthttpserver even more difficult is that the repo has zero documentation.

Have you try to run with "-qml" arguments ? If you don't do that never the file `webserverinfosender.cpp` is called in the default path of the program.

**Reply by ttonyh** [reply]
I haven't tried it, but will. Thank you.

**Reply by gungun974** [reply]
> If anyone here, however, would be very kind and patient to explain to me how with GDB I can find out where the program is segfaulting, I would be grateful for this help.

Okay I manage to understand and run GDB.

So here what happen : 
```
root@qt:~/qdomyos-zwift/src# gdb ./qdomyos-zwift
GNU gdb (Raspbian 10.1-1.7) 10.1.90.20210103-git
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "arm-linux-gnueabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./qdomyos-zwift...
(No debugging symbols found in ./qdomyos-zwift)
(gdb) run -no-gui -qml
Starting program: /root/qdomyos-zwift/src/qdomyos-zwift -no-gui -qml
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
OK, you are root.
[New Thread 0xf2a44380 (LWP 1947)]
[Detaching after fork from child process 1948]
[New Thread 0xf20ff380 (LWP 1950)]

Thread 3 "QQmlThread" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xf20ff380 (LWP 1950)]
0xf7033a4c in ?? () from /opt/Qt/5.15.9/lib/libQt5Gui.so.5
```

Well, I suppose I must have done something wrong when I compiled my version of QT because I suppose it's not normal for the program to try to access nothing in the QT graphic library. 

I'll keep you posted if I manage to solve this problem and progress in the quest to get this web server working xD.

**Reply by gungun974** [reply]
> > If anyone here, however, would be very kind and patient to explain to me how with GDB I can find out where the program is segfaulting, I would be grateful for this help.
> 
> Okay I manage to understand and run GDB.
> 
> So here what happen :
> 
> ```
> root@qt:~/qdomyos-zwift/src# gdb ./qdomyos-zwift
> GNU gdb (Raspbian 10.1-1.7) 10.1.90.20210103-git
> Copyright (C) 2021 Free Software Foundation, Inc.
> License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
> This is free software: you are free to change and redistribute it.
> There is NO WARRANTY, to the extent permitted by law.
> Type "show copying" and "show warranty" for details.
> This GDB was configured as "arm-linux-gnueabihf".
> Type "show configuration" for configuration details.
> For bug reporting instructions, please see:
> <https://www.gnu.org/software/gdb/bugs/>.
> Find the GDB manual and other documentation resources online at:
>     <http://www.gnu.org/software/gdb/documentation/>.
> 
> For help, type "help".
> Type "apropos word" to search for commands related to "word"...
> Reading symbols from ./qdomyos-zwift...
> (No debugging symbols found in ./qdomyos-zwift)
> (gdb) run -no-gui -qml
> Starting program: /root/qdomyos-zwift/src/qdomyos-zwift -no-gui -qml
> [Thread debugging using libthread_db enabled]
> Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
> OK, you are root.
> [New Thread 0xf2a44380 (LWP 1947)]
> [Detaching after fork from child process 1948]
> [New Thread 0xf20ff380 (LWP 1950)]
> 
> Thread 3 "QQmlThread" received signal SIGSEGV, Segmentation fault.
> [Switching to Thread 0xf20ff380 (LWP 1950)]
> 0xf7033a4c in ?? () from /opt/Qt/5.15.9/lib/libQt5Gui.so.5
> ```
> 
> Well, I suppose I must have done something wrong when I compiled my version of QT because I suppose it's not normal for the program to try to access nothing in the QT graphic library.
> 
> I'll keep you posted if I manage to solve this problem and progress in the quest to get this web server working xD.

Well, I tried to compile qdoymos directly without this HTTP module using the version included in my raspberry pi of QT 5.15.12 and I still have the same issue when I try to force QML with this time `/lib/arm-linux-gnueabihf/libQt5Gui.so.5 segment fault`.

I now assume that the problem must surely come from my Linux environment on my raspberry pi because it's not normal that I can't follow the basic Linux instructions of the project xD.

**Comment by gungun974** [comment]
Well, I finally succeeded on my side to actually make the Websocket module work by doing the opposite of what I wanted to do initially xD

I started over with a Raspberry Pi 4 using the Raspbian 64 bit Lite image.

I installed these packages:

```bash
sudo apt install git qtquickcontrols2-5-dev libqt5bluetooth5 libqt5widgets5 libqt5positioning5 libqt5xml5 qtconnectivity5-dev qtpositioning5-dev libqt5charts5-dev libqt5charts5 qt5-assistant libqt5networkauth5-dev libqt5websockets5-dev libqt5texttospeech5-dev libqt5texttospeech5 libqt5location5-plugins qtlocation5-dev qtmultimedia5-dev libqt5multimediawidgets5 libqt5multimedia5-plugins libqt5multimedia5 g++ make
```

I installed this package because I need it for my HACK that I'm going to do on qthttpserver:

```bash
sudo apt qtbase5-private-dev
```

I followed the instructions of my tutorial:

```bash
git clone https://github.com/qt-labs/qthttpserver
cd ~/qthttpserver/src/3rdparty/http-parser
wget https://raw.githubusercontent.com/nodejs/http-parser/main/http_parser.h
wget https://raw.githubusercontent.com/nodejs/http-parser/main/http_parser.c
cd ~/qthttpserver/src
```

Before BUILDING, I HACKED the httpserver.pro file to remove the "private" from websocket because why not try.

```bash
vim httpserver/httpserver.pro
```

I built the HTTP module and it works, strangely enough.

Next, I cloned the project `https://github.com/cagnulein/qdomyos-zwift.git` and tried to Build and I had no errors but it didn't work.

Then, I had the idea of the century. I will just compile the latest version that I know works for me:

```bash
git checkout 81001cbbdd3f64ca9b6ee50335eca71f4329a4d5.
```

I followed my tutorial to compile this old version.

```bash
cd ~/qdomyos-zwift
git submodule update --init ~/qdomyos-zwift/src/smtpclient/
cd ~/qdomyos-zwift/src
qmake
make -j4 # Remove -j4 if you compiled on Raspberry Pi Zero
```

I added this to my configuration, though:

```bash
template_user_QZWS_enabled=true
template_user_QZWS_folders=:/templates//debug, :/templates//example
template_user_QZWS_ips=192.168.1.154
template_user_QZWS_port=3005
template_user_QZWS_type=WebServer
```

I launched the program as root with -qml

I went to this URL: `http://192.168.1.154:3005/debug/workout.htm`

AND TADA it Works xD

![2023-05-25_16-28](https://github.com/cagnulein/qdomyos-zwift/assets/5224459/8cab49ad-3d41-444c-a863-15e59de55a45)

So, I don't know why I can't make it work in the latest version, but since I've been trying to get this project working for too long, I'm not asking myself any more questions than that.

**Reply by cagnulein** [reply]
great! I guess it can also work on the last version, i didn't change anything on that side. let's see how it goes for @ttonyh

**Reply by ttonyh** [reply]
@gungun974 Is it possible for you to try to get it to work on your system using the current master branch?

**Comment by ttonyh** [comment]
Thank you! I will try this in the next few days when I get the chance and report back.

**Comment by ttonyh** [comment]
@gungun974 @cagnulein I just tried a similar method, only I used the latest commit of QZ (not 81001cbbdd). I tried this method in a VM using VirtualBox with Debian 11 (Bullseye) and a mini x86 PC with Ubuntu 22.04.

Both the VM and the mini-PC seem to fist connect and look like they are working (but never with websockets) then give me a segmentation fault.

First my bike will beep and the connect light will go from blinking to solid blue. This indicates (from the bike's perspective) that a connection has been made.

Console output:


This...
```
Sat May 27 11:51:35 2023 1685213495502 Debug:   Discovered:  "ECHEX-3-189462" "C1:39:C1:9B:50:CB" Num UUIDs 6 total device 0 cached RSSI 0 Num Manufacture
rData 0
qt.bluetooth.bluez: Discovered:  "ECHEX-3-189462" "C1:39:C1:9B:50:CB" Num UUIDs 6 total device 0 cached RSSI 0 Num ManufacturerData 0
Sat May 27 11:51:35 2023 1685213495502 Debug: bluetooth.cpp void bluetooth::deviceDiscovered(const QBluetoothDeviceInfo&) manufacturerData
manufacturerData
Sat May 27 11:51:35 2023 1685213495502 Debug: bluetooth.cpp void bluetooth::debug(const QString&) "Found new device: ECHEX-3-189462 (C1:39:C1:9B:50:CB) \u
0000:\u0000"
"Found new device: ECHEX-3-189462 (C1:39:C1:9B:50:CB) \u0000:\u0000"
Sat May 27 11:51:35 2023 1685213495506 Debug:   void QBluetoothDeviceDiscoveryAgentPrivate::stop()
qt.bluetooth.bluez: void QBluetoothDeviceDiscoveryAgentPrivate::stop()
Sat May 27 11:51:35 2023 1685213495507 Debug: bluetooth.cpp void bluetooth::debug(const QString&) "BTLE scanning stops"
"BTLE scanning stops"
```


Then a bunch of these


```
Sat May 27 11:48:43 2023 1685213323719 Debug: virtualbike.cpp void virtualbike::reconnect() "virtualbike::reconnect"
"virtualbike::reconnect"
Sat May 27 11:48:43 2023 1685213323719 Warning:   Overriding existing local service with uuid "{00001826-0000-1000-8000-00805f9b34fb}"
Overriding existing local service with uuid "{00001826-0000-1000-8000-00805f9b34fb}"
Sat May 27 11:48:43 2023 1685213323719 Warning:   attribute of type "{00002902-0000-1000-8000-00805f9b34fb}" has invalid length of 3 bytes
qt.bluetooth.bluez: attribute of type "{00002902-0000-1000-8000-00805f9b34fb}" has invalid length of 3 bytes
Sat May 27 11:48:43 2023 1685213323719 Debug:   Starting to advertise
qt.bluetooth.bluez: Starting to advertise
Sat May 27 11:48:43 2023 1685213323719 Debug:   sending command; ogf: 8 ocf: 7
qt.bluetooth.bluez: sending command; ogf: 8 ocf: 7
Sat May 27 11:48:43 2023 1685213323719 Debug:   command sent successfully
qt.bluetooth.bluez: command sent successfully
Sat May 27 11:48:43 2023 1685213323720 Debug:   HCI event triggered, type: e
qt.bluetooth.bluez: HCI event triggered, type: e
Sat May 27 11:48:43 2023 1685213323720 Debug:   command 7 failed with status 12
qt.bluetooth.bluez: command 7 failed with status 12
Sat May 27 11:48:43 2023 1685213323720 Debug:   reading power level failed, leaving it out of the advertising data
qt.bluetooth.bluez: reading power level failed, leaving it out of the advertising data
Sat May 27 11:48:43 2023 1685213323720 Debug:   advertising parameters: "a000a0000000000000000000000700"
qt.bluetooth.bluez: advertising parameters: "a000a0000000000000000000000700"
Sat May 27 11:48:43 2023 1685213323720 Debug:   advertising data: "170201060d09446f6d796f734272696467650503261800ff0000000000000000"
qt.bluetooth.bluez: advertising data: "170201060d09446f6d796f734272696467650503261800ff0000000000000000"
Sat May 27 11:48:43 2023 1685213323721 Debug:   scan response data: "140d09446f6d796f734272696467650503261800ff0000000000000000000000"
qt.bluetooth.bluez: scan response data: "140d09446f6d796f734272696467650503261800ff0000000000000000000000"
Sat May 27 11:48:43 2023 1685213323721 Debug:   sending command; ogf: 8 ocf: 10
qt.bluetooth.bluez: sending command; ogf: 8 ocf: 10
Sat May 27 11:48:43 2023 1685213323721 Debug:   command sent successfully
qt.bluetooth.bluez: command sent successfully
Sat May 27 11:48:43 2023 1685213323722 Debug:   HCI event triggered, type: e
qt.bluetooth.bluez: HCI event triggered, type: e
Sat May 27 11:48:43 2023 1685213323722 Debug:   command 10 failed with status 12
qt.bluetooth.bluez: command 10 failed with status 12
Sat May 27 11:48:43 2023 1685213323722 Debug:   Advertising disable failed, ignoring
qt.bluetooth.bluez: Advertising disable failed, ignoring
Sat May 27 11:48:43 2023 1685213323722 Debug:   sending command; ogf: 8 ocf: 6
qt.bluetooth.bluez: sending command; ogf: 8 ocf: 6
Sat May 27 11:48:43 2023 1685213323722 Debug:   command sent successfully
qt.bluetooth.bluez: command sent successfully
Sat May 27 11:48:43 2023 1685213323723 Debug:   HCI event triggered, type: e
qt.bluetooth.bluez: HCI event triggered, type: e
```


Then a bunch of these

```
Sat May 27 11:54:46 2023 1685213686642 Debug:   command sent successfully
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: command sent successfully
Sat May 27 11:54:46 2023 1685213686642 Debug:   sending command; ogf: 8 ocf: 7
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: sending command; ogf: 8 ocf: 7
Sat May 27 11:54:46 2023 1685213686642 Debug:   command sent successfully
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: command sent successfully
Sat May 27 11:54:46 2023 1685213686642 Warning:   Unexpected closure of device. Cleaning up internal states.
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: Unexpected closure of device. Cleaning up internal states.
Sat May 27 11:54:46 2023 1685213686645 Debug: virtualbike.cpp void virtualbike::reconnect() "virtualbike::reconnect"
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
"virtualbike::reconnect"
Sat May 27 11:54:46 2023 1685213686645 Warning:   attribute of type "{00002902-0000-1000-8000-00805f9b34fb}" has invalid length of 3 bytes
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: attribute of type "{00002902-0000-1000-8000-00805f9b34fb}" has invalid length of 3 bytes
Sat May 27 11:54:46 2023 1685213686645 Debug:   Starting to advertise
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: Starting to advertise
Sat May 27 11:54:46 2023 1685213686645 Debug:   sending command; ogf: 8 ocf: 7
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: sending command; ogf: 8 ocf: 7
Sat May 27 11:54:46 2023 1685213686646 Debug:   command sent successfully
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: command sent successfully
Sat May 27 11:54:46 2023 1685213686646 Warning:   socket creation failed: "Too many open files"
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: socket creation failed: "Too many open files"
Sat May 27 11:54:46 2023 1685213686646 Debug: virtualbike.cpp void virtualbike::error(QLowEnergyController::Error) "virtualbike::controller:ERROR Advertis
ingError"
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
"virtualbike::controller:ERROR AdvertisingError"
Sat May 27 11:54:46 2023 1685213686646 Debug: virtualbike.cpp void virtualbike::reconnect() "virtualbike::reconnect"
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
"virtualbike::reconnect"
Sat May 27 11:54:46 2023 1685213686646 Debug:   sending command; ogf: 8 ocf: 7
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: sending command; ogf: 8 ocf: 7
Sat May 27 11:54:46 2023 1685213686646 Debug:   command sent successfully
QIODevice::write (QFile, "debug-Sat_May_27_11_54_07_2023.log"): device not open
qt.bluetooth.bluez: command sent successfully
```


Then finally a segmentation fault

```
Sat May 27 11:54:46 2023 1685213686650 Warning:   attribute of type "{00002902-0000-1000-8000-00805f9b34fb}" has invalid length of 3 bytes
Segmentation fault (core dumped)
```

**Reply by cagnulein** [reply]
@ttonyh can you try disabling the virtual device setting ?

**Reply by ttonyh** [reply]
Yes I think this setting did it:
```virtual_device_bluetooth=false```

But still no websockets.

Here's my WS config:

```
template_user_QZWS_enabled=true
template_user_QZWS_folders=:/inner_templates//chartjs
template_user_QZWS_ips=0.0.0.0
template_user_QZWS_port=9999
template_user_QZWS_type=WebServer
```

Should I be using ```template_inner_``` or ```template_user_``` or it doesn't matter?

**Reply by cagnulein** [reply]
did you still have segmentation fault now?

these are  classic setup for qz

```
Tue May 23 10:15:23 2023 1684862123460 Debug: ../src/main.cpp int main(int, char **) "template_inner_QZWS_enabled" QVariant(QString, "true")
Tue May 23 10:15:23 2023 1684862123461 Debug: ../src/main.cpp int main(int, char **) "template_inner_QZWS_folders" QVariant(QStringList, (":/inner_templates//floating", ":/inner_templates//maps2d", ":/inner_templates//chartjs", ":/inner_templates//googlemaps"))
Tue May 23 10:15:23 2023 1684862123462 Debug: ../src/main.cpp int main(int, char **) "template_inner_QZWS_ips" QVariant(Invalid)
Tue May 23 10:15:23 2023 1684862123463 Debug: ../src/main.cpp int main(int, char **) "template_inner_QZWS_port" QVariant(QString, "43397")
Tue May 23 10:15:23 2023 1684862123464 Debug: ../src/main.cpp int main(int, char **) "template_inner_QZWS_type" QVariant(QString, "WebServer")
Tue May 23 10:15:23 2023 1684862123465 Debug: ../src/main.cpp int main(int, char **) "template_inner_ids" QVariant(QString, "inner_QZWS")
Tue May 23 10:15:23 2023 1684862123466 Debug: ../src/main.cpp int main(int, char **) "template_user_QZWS_enabled" QVariant(QString, "false")
Tue May 23 10:15:23 2023 1684862123466 Debug: ../src/main.cpp int main(int, char **) "template_user_QZWS_folders" QVariant(QStringList, (":/templates//debug", ":/templates//example"))
Tue May 23 10:15:23 2023 1684862123467 Debug: ../src/main.cpp int main(int, char **) "template_user_QZWS_ips" QVariant(Invalid)
Tue May 23 10:15:23 2023 1684862123468 Debug: ../src/main.cpp int main(int, char **) "template_user_QZWS_type" QVariant(QString, "WebServer")
Tue May 23 10:15:23 2023 1684862123469 Debug: ../src/main.cpp int main(int, char **) "template_user_ids" QVariant(QStringList, ("user_QZWS", "user_qz", "user_vlc"))
Tue May 23 10:15:23 2023 1684862123470 Debug: ../src/main.cpp int main(int, char **) "template_user_qz_enabled" QVariant(QString, "false")
Tue May 23 10:15:23 2023 1684862123471 Debug: ../src/main.cpp int main(int, char **) "template_user_qz_type" QVariant(QString, "TcpClient")
Tue May 23 10:15:23 2023 1684862123472 Debug: ../src/main.cpp int main(int, char **) "template_user_vlc_enabled" QVariant(QString, "false")
Tue May 23 10:15:23 2023 1684862123473 Debug: ../src/main.cpp int main(int, char **) "template_user_vlc_type" QVariant(QString, "TcpClient")
```

**Reply by ttonyh** [reply]
No seg fault. It's working well, I can see output changes when the bike moves etc.:

```
Sat May 27 12:33:19 2023 1685215999832 Debug: echelonconnectsport.cpp void echelonconnectsport::characteristicChanged(const QLowEnergyCharacteristic&, con
st QByteArray&) "Last CrankEventTime: 0"
"Last CrankEventTime: 0"
Sat May 27 12:33:19 2023 1685215999832 Debug: echelonconnectsport.cpp void echelonconnectsport::characteristicChanged(const QLowEnergyCharacteristic&, con
st QByteArray&) "Current Watt: 0"
"Current Watt: 0"
Sat May 27 12:33:20 2023 1685216000818 Debug: echelonconnectsport.cpp void echelonconnectsport::characteristicChanged(const QLowEnergyCharacteristic&, con
st QByteArray&) " << f0 d1 09 0c 2b 00 00 00 0f 00 00 00 10"
" << f0 d1 09 0c 2b 00 00 00 0f 00 00 00 10"
Sat May 27 12:33:20 2023 1685216000820 Debug: echelonconnectsport.cpp void echelonconnectsport::characteristicChanged(const QLowEnergyCharacteristic&, con
st QByteArray&) "Current Local elapsed: 00:51:55"
"Current Local elapsed: 00:51:55"
Sat May 27 12:33:20 2023 1685216000820 Debug: echelonconnectsport.cpp void echelonconnectsport::characteristicChanged(const QLowEnergyCharacteristic&, con
st QByteArray&) "Current Speed: 0"
"Current Speed: 0"
Sat May 27 12:33:20 2023 1685216000820 Debug: echelonconnectsport.cpp void echelonconnectsport::characteristicChanged(const QLowEnergyCharacteristic&, con
st QByteArray&) "Current Calculate Distance: 0"

```

By the way, a side-note. I purchased and have been using the iOS app and it works well. Thank you!

**Reply by cagnulein** [reply]
Ok so in the log don't you see the template sender lines? If not it means that the web socket is not activated in the qt build

**Reply by ttonyh** [reply]
I don't see anything about template sender in the output.

**Reply by cagnulein** [reply]
Ok so for sure the mkspecs files aren't in the right place

**Reply by ttonyh** [reply]
What do you suggest or how do I troubleshoot this?

**Reply by cagnulein** [reply]
Double check the info here https://github.com/cagnulein/qdomyos-zwift/pull/252

**Reply by cagnulein** [reply]
Going to bed now:)

**Reply by ttonyh** [reply]
To clarify, I don't even get an open http port when I run QZ.

**Comment by cagnulein** [comment]
Yes that’s exactly the issue. Web sockets seems disabled by compilation on
your setup. Check the link that I sent.

Il giorno sab 27 mag 2023 alle 21:53 Tony Habash ***@***.***>
ha scritto:

> To clarify, I don't even get an open http port when I run QZ.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cagnulein/qdomyos-zwift/discussions/1461#discussioncomment-6020360>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALYWFPCPNWYZSQPQ35XY3XIJLUZANCNFSM6AAAAAAYAUTYGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>
-- 
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud

**Reply by ttonyh** [reply]
@cagnulein Have you tried to get websockets working on the latest QZ source in Linux? I noticed that qthttpserver is only tested in the Windows build (in github actions workflow). I just tried to check out 81001cb and compile/run as noted above and that was the first time I was able to see the websocket port open on the computer. However I couldn't get any output from it. Also I didn't understand the solution that you mentioned regarding #1461.

**Reply by cagnulein** [reply]
let me setup into the CI as I did for the windows version

**Reply by cagnulein** [reply]
done, let's see if works https://github.com/cagnulein/qdomyos-zwift/pull/1489

**Reply by cagnulein** [reply]
@ttonyh let me know when you will try it

**Reply by ttonyh** [reply]
I forked the repo and ran the action. Builds just fine. Thanks for adding that. Have you been able to get websockets working on your local environment using the same method as the Github Action in Linux?

**Comment by ttonyh** [comment]
As soon as I get this to work, I'm happy to open a PR and thoroughly document this procedure.

**Reply by cagnulein** [reply]
i didn't tried, i don't have time in these days. my main job is killing me and in the weekend i'm trying to save time with my family. let me know how it goes

**Comment by cagnulein** [comment]
also you can use the  binary from the github actions directly now

Il giorno lun 29 mag 2023 alle 04:14 Tony Habash ***@***.***>
ha scritto:

> I forked the repo and ran the action. Builds just fine. Thanks for adding
> that. Have you been able to get websockets working on your local
> environment using the same method as the Github Action in Linux?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cagnulein/qdomyos-zwift/discussions/1461#discussioncomment-6025777>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALYWDRPO2BWVVXMLAIXA3XIQA77ANCNFSM6AAAAAAYAUTYGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>
-- 
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud
